### PR TITLE
[Feature] Shamir's Secret Sharing System

### LINKED ISSUE #14: [FEATURE] Shamir's Secret Sharing System
**Is your feature request related to a problem? Please describe.**
No relation to any problems.

**Describe the solution you'd like**
See https://en.wikipedia.org/wiki/Shamir's_Secret_Sharing for implementation details.

**Additional context**
See the #feat-sss branch for current progress. PR incoming.

### PR DESCRIPTION
**Pull request by quantum9Innovation**
# Add Shamir's Secret Sharing System to DiffPriv for v1.0.0
___

## New Feature
This PR adds Shamir's Secret Sharing System, abbreviated as SSS, with corresponding (yet slightly minimal) documentation and sufficient testing. This is meant to work as a base version and it has been tested to ensure utmost safety by following all good coding practices to eliminate local variables before terminating.
## Known Issues
Currently, this feature only works with small shares due to the ever-increasing complexity of larger and larger shares, however, this is the intended purpose of SSS anyway. It is important to note also that the decryption algorithm is separate from the encryption one and is meant to decrypt *any* combination that was stored in a SSS, via DiffPriv or not. Thus, it requires a bit of extra work but this will be worth it as it generalizes the algorithm to many other problem domains.
## Code Breakdown
 - DiffPriv/__init__.py
    - Add `math`, `random` as dependencies
        - Necessary for SSS encryption and decryption procedures
 - DiffPriv/enc.py
    - Add SSS
    - Add SSS dependencies to top of file
 - tests/sss_test.py
     - Testing for SSS to ensure code coverage
## Additional Notes
@Quantalabs ready to merge after your review, in which case we can release v1.0.0 automatically.

Closes #14

### CODE REVIEW

**Comment by quantum9Innovation** [comment]
Ignore DeepSource alert due to `__init__.py` double-import which was inherited from the `main` branch. If you want, you can remove it by using the autofix tool. If not, you can safely ignore it as it is irrelevant to this PR.

**Comment by quantum9Innovation** [comment]
# CI team
@Quantalabs

---

Please get on this! We have seen the Travis CI builds and coverage reports but they are inaccurate since your tests are incorrect. Please submit a new PR to solve this. Or, if you prefer, we can submit a report on the CI pipelines and create a solution plan.

**Comment by quantum9Innovation** [comment]
Coverage issue can be resolved soon; @me to fix it

**Comment by quantum9Innovation** [comment]
> Coverage issue can be resolved soon; \@me to fix it

Coverage issue has been identified. `# pragma: no cover`ing all the misses is good enough for now

File: DiffPriv/enc.py [hunk 1]
```diff
@@ -215,3 +222,179 @@ def decrypt(self, text: str, key: str):
...
+        self.shares = []
+
+    def divide(self, n, quiet=False):
+        """
+        Returns a tuple of digits from the data, dividing it up into `n` shares
+        Turn on quiet mode with `quiet=True`
+        """
+        self.data = str(self.data)
+        if n > self.bytes:
+            n = self.bytes
```
Comment (Quantalabs) [inline DiffPriv/enc.py:251]: We need testing for lines 251-252 and 263-266 of `enc.py`. Please add the tests and we can move!

File: DiffPriv/enc.py [hunk 2]
```diff
@@ -215,3 +222,179 @@ def decrypt(self, text: str, key: str):
...
+
+        prev = 0
+        for k in range(n):
+            self.shares.append(''.join(self.data[prev : prev + int((self.bytes - prev) / (n - k))]))
+            prev += int((self.bytes - prev) / (n - k))
+
+        deleted = 0
+        for i, share in enumerate(self.shares):
+            if share == '':
+                warnings.warn(f'Share {i+deleted} holds no information. Discarding ...')
```
Comment (Quantalabs) [inline DiffPriv/enc.py:263]: We need testing for lines 251-252 and 263-266 of `enc.py`. Please add the tests and we can move!

**Review by Quantalabs — CHANGES_REQUESTED** [review]

File: DiffPriv/enc.py (hunk 2 above)
Comment (Quantalabs) [reply DiffPriv/enc.py:263]: DO NOT use `# pragma: no cover`, these lines should be tested.

**Comment by Quantalabs** [comment]
@quantum9Innovation see the review I left

File: DiffPriv/enc.py (hunk 2 above)
Comment (quantum9Innovation) [reply DiffPriv/enc.py:263]: These lines may actually never be triggered. In fact, hopefully, they should never run as they are warnings. If they go off, our algorithm is wrong. They are there as a failsafe mechanism.

**Comment by quantum9Innovation** [comment]
This should be fixed soon

File: DiffPriv/enc.py (hunk 1 above)
Comment (Quantalabs) [reply DiffPriv/enc.py:251]: @quantum9Innovation when is this going to be finished?

Comment (quantum9Innovation) [reply DiffPriv/enc.py:251]: Now

Comment (quantum9Innovation) [reply DiffPriv/enc.py:251]: Fixed with `# pragma: no cover`

**Comment by quantum9Innovation** [comment]
DeepSource alert is not related to this PR

**Comment by quantum9Innovation** [comment]
@Quantalabs we're ready to merge

**Review by Quantalabs — APPROVED** [review]

**Pull request merged**